backend/sns/users/controller.py

```diff
@@ -1,8 +1,287 @@
-from fastapi import APIRouter
+from fastapi import APIRouter, Depends, status, HTTPException, Body
+from starlette.background import BackgroundTasks
+from sqlalchemy.orm import Session
 
-router = APIRouter(prefix="/users")
+from sns.common.session import db
+from sns.users.service import user_service
+from sns.users.schema import (
+    UserPasswordUpdate,
+    UserCreate,
+    UserUpdate,
+    UserBase,
+    UserRead,
+    Token,
+    Msg,
+)
 
 
-@router.get("/")
-def root():
-    return {"message": "Hello world"}
+router = APIRouter()
+
+
+@router.post("/signup", response_model=Msg, status_code=status.HTTP_201_CREATED)
+def signup(
+    data_for_signup: UserCreate,
+    background_tasks: BackgroundTasks,
+    db: Session = Depends(db.get_db),
+):
+    """email과 password로 새 user를 등록한다.
+
+    Args:
+
+    - data_for_signup (schema.UserCreate) : 등록할 email과 password 정보
+
+    Raises:
+
+    - HTTPException (403 FORBIDDEN): 다음 2가지 경우에 발생한다.
+        - 회원 가입 시 입력한 이메일이 이미 인증되었을 때
+        - 이미 등록은 되었고, 인증은 안되었을 때
+    - HTTPException (500 INTERNAL SERVER ERROR): 다음 2가지 경우에 발생한다.
+        - 유저 생성에 실패했을 때
+        - 이메일 인증을 위한 이메일 발송에 실패했을 때
+
+
+    Returns:
+
+    - Msg: 이메일 전송 성공 유무 메세지 반환
+    """
+    user = user_service.get_user(db, email=data_for_signup.email)
+
+    if user:
+        if user_service.is_verified(user):
+            raise HTTPException(status_code=403, detail="이미 인증된 이메일입니다.")
+        else:
+            raise HTTPException(status_code=403, detail="이미 등록되었고, 미인증인 유저입니다.")
+
+    user_service.create(db, data_for_signup)  # 미인증 유저 생성
+
+    # 이메일 인증 메일 발송하기
+    user_service.send_verification_email(
+        db, email=data_for_signup.email, background_tasks=background_tasks
+    )
+    return {"status": "success", "msg": "이메일 전송이 완료되었습니다."}
+
+
+@router.patch(
+    "/verification-email/{code}", response_model=Msg, status_code=status.HTTP_200_OK
+)
+def verify_email(code: str, db: Session = Depends(db.get_db)):
+    """code 정보를 받아 user를 조회하여 해당 user의 인증 상태를 True로 바꾼다.
+
+    Args:
+
+    - code (str) : url에 담겨진 인증 code 정보
+
+    Raises:
+
+    - HTTPException (404 NOT FOUND): 다음 경우에 발생한다.
+        - verification code가 code 값과 일치하는 user를 찾지 못할 때 발생한다.
+    - HTTPException (500 INTERNAL SERVER ERROR): 인증 상태값 변경에 실패했을 때 발생한다.
+
+    Returns:
+
+    - Msg: 계정 인증 완료 메세지
+    """
+    status.HTTP_404_NOT_FOUND
+    user = user_service.get_user(db, verification_code=code)
+    user_service.update(db, user=user, data_to_be_updated={"verified": True})
+    return {"status": "success", "msg": "이메일 인증이 완료되었습니다."}
+
+
+@router.post("/login", response_model=Token, status_code=status.HTTP_200_OK)
+def login(
+    email: str = Body(...), password: str = Body(...), db: Session = Depends(db.get_db)
```

Comment:
pre commit config yaml 에 https://github.com/asottile/add-trailing-comma 추가하면, 아래와 같이 바꿔줍니다. 

```suggestion
    email: str = Body(...), password: str = Body(...), db: Session = Depends(db.get_db),
```

그리고 여기에 black 을 적용하면 이렇게 됩니다

```suggestion
    email: str = Body(...), 
    password: str = Body(...), 
    db: Session = Depends(db.get_db),
```

이렇게 하는 것이 좋습니다. 

https://peps.python.org/pep-0008/#when-to-use-trailing-commas

인자가 추가될 수 있을 때, 곧 리스트나 튜플같은 형태가 여러 줄로 나열되었을 땐 매 라인 끝에 comma 를 붙여주는 것이 좋습니다. 

- 새로운 인자를 추가하거나 삭제할 때, 훨씬 간편합니다.